### PR TITLE
[JS BE] ES6-classes

### PR DESCRIPTION
**Pull request by tihonovcore**

### CODE REVIEW

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/ObjectLowering.kt [hunk 1]
```diff
@@ -87,7 +84,19 @@ class ObjectUsageLowering(
                 val initInstanceField = context.createIrBuilder(container.symbol).buildStatement(UNDEFINED_OFFSET, UNDEFINED_OFFSET) {
                     irSetField(null, instanceField, irGet(irClass.thisReceiver!!))
                 }
-                (irBody as IrBlockBody).statements.add(0, initInstanceField)
+                if (context.es6mode) {
+                    //find super and put after
+                    val index = (irBody as IrBlockBody).statements
+                        .indexOfFirst { it is IrDelegatingConstructorCall || it is IrVariable && it.origin === ES6_THIS_VARIABLE_ORIGIN }
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/ObjectLowering.kt:90]: `IrElementTransformer`?

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/SecondaryCtorLowering.kt [hunk 2]
```diff
@@ -240,16 +244,24 @@ private class CallsiteRedirectionTransformer(private val context: JsIrBackendCon
...
-                    IrGetValueImpl(startOffset, endOffset, thisReceiver.type, thisReceiver.symbol)
-                } else {
-                    val lastValueParameter = data.valueParameters.last()
-                    IrGetValueImpl(startOffset, endOffset, lastValueParameter.type, lastValueParameter.symbol)
+            val readThis: IrExpression
+            if (context.es6mode) {
+                return newCall
+            } else {
+                readThis = expression.run {
+                    if (data!! is IrConstructor) {
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/SecondaryCtorLowering.kt:258]: `data is IrConstructor` ?

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/ThrowableLowering.kt [hunk 3]
```diff
@@ -24,7 +24,7 @@ class ThrowableLowering(val context: JsIrBackendContext) : BodyLoweringPass {
 
     private val throwableConstructors = context.throwableConstructors
     private val newThrowableFunction = context.newThrowableSymbol
-    private val extendThrowableFunction = context.extendThrowableSymbol
+    private val extendThrowableFunction = with(context) { if (es6mode) setPropertiesToThrowableInstanceSymbol else extendThrowableSymbol }
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/ThrowableLowering.kt:27]: Let pass it to constructor

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/transformers/irToJs/JsClassGenerator.kt [hunk 4]
```diff
@@ -111,18 +139,40 @@ class JsClassGenerator(private val irClass: IrClass, val context: JsGenerationCo
...
 
             return jsAssignment(memberRef, translatedFunction.apply { name = null }).makeStmt()
         }
 
+        generateOverrides(declaration)
+
+        return null
+    }
+
+    private fun generateMemberFunctionES6(declaration: IrSimpleFunction): JsFunction? {
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/transformers/irToJs/JsClassGenerator.kt:157]: Seems like both functions could be merged

File: js/js.tests/build.gradle.kts [hunk 5]
```diff
@@ -145,6 +146,25 @@ projectTest("jsIrTest", true) {
...
+projectTest("jsEs6IrTest", true) {
+    systemProperty("kotlin.js.ir.pir", "false")
+    systemProperty("kotlin.js.ir.es6", "true")
+
+    dependsOn(":dist")
+    dependsOn(":kotlin-stdlib-js-ir:generateFullRuntimeKLib")
+    dependsOn(":kotlin-stdlib-js-ir:generateReducedRuntimeKLib")
+    dependsOn(":kotlin-stdlib-js-ir:generateKotlinTestKLib")
+
+    exclude("org/jetbrains/kotlin/js/test/wasm/semantics/*")
```
Comment (romanart) [inline js/js.tests/build.gradle.kts:158]: Why?

File: js/js.tests/build.gradle.kts [hunk 6]
```diff
@@ -145,6 +146,25 @@ projectTest("jsIrTest", true) {
...
+    exclude("org/jetbrains/kotlin/js/test/ir/semantics/*")
```
Comment (romanart) [inline js/js.tests/build.gradle.kts:159]: Why?

Comment (tihonovcore) [reply js/js.tests/build.gradle.kts:159]: It's runs tests with TargetBackend.JS_IR
Ir tests with TargetBackend.JS_IR_ES6 here: `org/jetbrains/kotlin/js/test/es6/semantics/*`

File: js/js.tests/test/org/jetbrains/kotlin/js/test/es6/semantics/abstractClassesForGeneratedIrTests.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import org.jetbrains.kotlin.test.TargetBackend
+
+abstract class AbstractIrBoxJsES6Test : BasicIrBoxES6Test(TEST_DATA_DIR_PATH + "box/", "irBox/")
+
+abstract class AbstractIrJsCodegenBoxES6Test : BasicIrBoxES6Test(
+    "compiler/testData/codegen/box/",
+    "codegen/irBox/"
+)
+
+abstract class AbstractIrWasmBoxJsES6Test : BasicIrBoxES6Test(
```
Comment (romanart) [inline js/js.tests/test/org/jetbrains/kotlin/js/test/es6/semantics/abstractClassesForGeneratedIrTests.kt:19]: This is not actually needed

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/EnumClassLowering.kt [hunk 8]
```diff
@@ -362,7 +362,12 @@ class EnumEntryInstancesBodyLowering(val context: JsIrBackendContext) : BodyLowe
             val enum = entryClass.parentAsClass
             if (enum.goodEnum) {
                 val entry = enum.declarations.filterIsInstance<IrEnumEntry>().find { it.correspondingClass === entryClass }!!
-                (irBody as IrBlockBody).statements.add(0, context.createIrBuilder(container.symbol).run {
+
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/EnumClassLowering.kt:365]: This code worth some explanation.

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/Dce.kt [hunk 9]
```diff
@@ -320,6 +325,13 @@ fun usefulDeclarations(roots: Iterable<IrDeclaration>, context: JsIrBackendConte
                                 toStringMethod.enqueue("intrinsic: jsPlus")
                             }
                         }
+                        context.intrinsics.jsConstruct -> {
+                            val arg = expression.getValueArgument(0)!!
+                            val jsClassCall = (arg as? IrCall) ?: ((arg as IrGetValue).symbol.owner as IrVariable).initializer
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/Dce.kt:330]: Unclear

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/Dce.kt [hunk 10]
```diff
@@ -296,6 +295,12 @@ fun usefulDeclarations(roots: Iterable<IrDeclaration>, context: JsIrBackendConte
                             val ref = expression.getTypeArgument(0)!!.classifierOrFail.owner as IrDeclaration
                             ref.enqueue("intrinsic: jsClass")
                             referencedJsClasses += ref
+
+                            //HACK: es6 result type - construct class in case of
+                            //`val currResultType = resultType || klass`
+                            val arg = expression.getTypeArgument(0)!!
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/Dce.kt:301]: I would suggest to introduce an additional intrinsic and track it separately

File: compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/JsLoweringPhases.kt [hunk 11]
```diff
@@ -228,7 +228,16 @@ private val removeInlineFunctionsWithReifiedTypeParametersLoweringPhase = makeDe
 )
 
 private val throwableSuccessorsLoweringPhase = makeBodyLoweringPhase(
-    ::ThrowableLowering,
+    { context ->
+        val extendThrowableSymbol = with(context) {
```
Comment (romanart) [inline compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/JsLoweringPhases.kt:232]: `context.run { .. }`

File: js/js.translator/testData/box/rtti/prototypeUsedToFindInterface.kt [hunk 12]
```diff
@@ -1,4 +1,7 @@
 // EXPECTED_REACHABLE_NODES: 1371
+// IGNORE_BACKEND: JS_IR_ES6
+// IGNORE_BACKEND: JS_IR_ES6
+
```
Comment (romanart) [inline js/js.translator/testData/box/rtti/prototypeUsedToFindInterface.kt:4]: x2

**Comment by romanart** [comment]
Merged manually

**Pull request closed without merging**